src/olmo_core/optim/skip_step_optimizer.py

```diff
@@ -0,0 +1,90 @@
+    interval.
+
+    .. important::
+        When using a :class:`SkipStepOptimizer` you must always set :data:`latest_loss` and
+        :data:`latest_grad_norm` to the current loss and grad norm, respectively, *before* calling
+        :meth:`step()`.
+
+        The :class:`~olmo_core.train.Trainer` will automatically set the :data:`latest_loss` whenever
+        its optimizer is a subclass of :class:`SkipStepOptimizer`, and the
+        :class:`~olmo_core.train.callbacks.GradClipperCallback` will automatically set the
+        :data:`latest_grad_norm`.
+
+    .. tip::
+        When implementing a :class:`SkipStepOptimizer` you should be careful to avoid host-device
+        syncs. You can use :meth:`get_step_factor()` within your :meth:`step()` method to do this.
+        See the implementation of :class:`SkipStepLion` for an example.
+    """
+
+    def __init__(
+        self,
+        params: ParamsT,
+        defaults: Dict[str, Any],
+        rolling_interval_length: int = 128,
+        sigma_factor: int = 6,
+    ) -> None:
+        super().__init__(params, defaults)
+        self.rolling_interval_length = rolling_interval_length
+        self.sigma_factor = sigma_factor
+        self._losses: List[torch.Tensor] = []
+        self._grad_norms: List[torch.Tensor] = []
+
+    @property
+    def latest_loss(self) -> torch.Tensor:
+        if not self._losses:
+            raise RuntimeError("'latest_loss' has not been set yet")
+        return self._losses[-1]
+
+    @latest_loss.setter
+    def latest_loss(self, loss: torch.Tensor):
+        self._losses.append(loss)
+        while len(self._losses) > self.rolling_interval_length + 1:
+            self._losses.pop(0)
+
+    @property
+    def latest_grad_norm(self) -> Optional[torch.Tensor]:
+        if not self._grad_norms:
+            return None
+        else:
+            return self._grad_norms[-1]
+
+    @latest_grad_norm.setter
+    def latest_grad_norm(self, grad_norm: torch.Tensor):
+        self._grad_norms.append(grad_norm)
+        while len(self._grad_norms) > self.rolling_interval_length + 1:
+            self._grad_norms.pop(0)
+
+    def get_step_factor(self) -> torch.Tensor:
+        """
+        Returns a float tensor which will be `1.0` if the optimizer should proceed with the step
+        and `0.0` if the optimizer should skip the step.
+
+        The tensor can be used within the optimizer's step computation to essentially skip a step
+        without a host-device sync.
+        """
+        if len(self._losses) < max(20, self.rolling_interval_length // 2):
+            return torch.tensor(1.0).to(device=self.latest_loss.device, non_blocking=True)
+
+        loss_std = torch.std(torch.stack(self._losses[:-1]))
+
+        if self._grad_norms:
+            grad_norm_std = torch.std(torch.stack(self._grad_norms[:-1]))
+            return (
+                self.latest_loss <= self.sigma_factor * loss_std
```

Comment:
The Karpathy thing actually filters out when loss/grad norm is `sigma_factor` standard deviations above the mean (this does above 0). A number that is `s` standard deviations above the mean is said to have a "z-score" of `s`.

Reply:
Oh, duh, that's embarrassing after having spent 4 years in a stats PhD program

Reply:
82d2c52

Reply:
Forgot the `.abs()` but just pushed a fix for that.